src/lib/vault/btc-provider.js

```diff
@@ -0,0 +1,183 @@
+
+    // This means we will send all the funds to the destination
+    const output = [{ address: to }];
+
+    try {
+      const result = coinSelection(utxos, output, Number(feeRate));
+      const { inputs, outputs } = result;
+
+      const psbt = this.createPsbt(inputs, outputs, descriptor);
+
+      return psbt.toHex();
+    } catch (error) {
+      throw new Error(error.message);
+    }
+  }
+
+  deriveAddress(publicKey) {
+    return bitcoin.payments.p2wpkh({ network: this.network, pubkey: Buffer.from(publicKey, 'hex') }).address;
+  }
+
+  deriveMultisigAddress(descriptor) {
+    const redeem = this.getRedeemScriptFromDescriptor(descriptor);
+    const { address } = bitcoin.payments.p2wsh({ network: this.network, redeem });
+
+    return address;
+  }
+
+  async getAddressBalance(address) {
+    const { confirmed, unconfirmed } = await this.instance.getAddressBalance(address);
+
+    if (unconfirmed < 0) {
+      throw new Error('TransactionPending');
+    }
+
+    if (confirmed <= 0) {
+      throw new Error('AccountNotFound');
+    }
+
+    return satsToBtc(confirmed);
+  }
+  async getFee() {
+    const fee = await this.instance.estimateFee(BTCFEERATE);
+
+    return satsToBtc(fee);
+  }
+
+  getParticipantsFromDescriptor(descriptor) {
+    const format = new RegExp(/^.sh\(sortedmulti\(2,(.*?)\)\)/);
+    const [, multisigString] = descriptor.match(format);
+    const [firstKey, ...keyStrings] = multisigString.split(',');
+    const signers = keyStrings.map(keyString => {
+      const [, pathString] = keyString.match(/\[(.*?)\]/);
+      const [fingerprint, ...paths] = pathString.split('/');
+      const [, publicKey] = keyString.split(']');
+
+      return {
+        fingerprint,
+        path: `m/${paths.join('/')}`,
+        publicKey
+      };
+    });
+
+    return [{ publicKey: firstKey }, ...signers];
+  }
+
+  getRedeemScriptFromDescriptor(descriptor) {
+    const participants = this.getParticipantsFromDescriptor(descriptor);
+
+    return getRedeemScript(this.network, participants, threshold);
+  }
+
+  getSignersFromDescriptor(descriptor) {
+    const participants = this.getParticipantsFromDescriptor(descriptor);
+
+    return participants.map(participant => this.deriveAddress(participant.publicKey));
+  }
+
+  get isConnected() {
+    return this.instance.isConnected();
+  }
+
+  multisigner(tx, keys) {
+    let signedTx = tx;
+
+    keys.forEach(key => {
+      signedTx = this.signer(key, signedTx);
+    });
+
+    return signedTx;
+  }
+
+  async sendTransaction(encodedTransaction, { to, amount }) {
+    const psbt = bitcoin.Psbt.fromHex(encodedTransaction);
+
+    if (!psbt.validateSignaturesOfAllInputs(signatureValidator)) {
+      throw new Error('InvalidSignerError');
+    }
+
+    psbt.finalizeAllInputs();
+
+    const tx = psbt.extractTransaction();
+
+    await this.instance.broadcastTransaction(tx.toHex());
+
+    return { amount, from: '', hash: tx.getId(), network: blockchain, to };
+  }
+
+  signer(key, tx) {
```

Comment:
```suggestion
  signTransaction(key, tx) {
```

Reply:
and since it will always be multiple keys you can remove multisigner and move logic here maybe

Reply:
👍 Great call, refactored to make it simple.